### ISSUE TITLE
Focus Window does not seem to function consistently

### ISSUE THREAD

**Issue by LexiconCode** [dictation-toolbox/dragonfly#265]
```
'switch to (vs|code)':
            FocusWindow(executable="Code.exe"),
```

```
Navigation: switch to (vs|code)
action.exec (ERROR): Execution of FocusWindow(executable='Code.exe') failed due to exception: (0, 'SetForegroundWindow', 'No error message is available')
Traceback (most recent call last):
  File "C:\Python27\lib\site-packages\dragonfly\actions\action_base.py", line 98, in execute
    if self._execute(data) is False:
  File "C:\Python27\lib\site-packages\dragonfly\actions\action_focuswindow.py", line 100, in _execute
    window.set_foreground()
  File "C:\Python27\lib\site-packages\dragonfly\windows\win32_window.py", line 244, in set_foreground
    self._set_foreground()
  File "C:\Python27\lib\site-packages\dragonfly\windows\win32_window.py", line 131, in <lambda>
    return lambda self: func(self._handle)
error: (0, 'SetForegroundWindow', 'No error message is available')
```

**Comment by tieTYT** [comment]
this is my issue that I am experiencing. Don't really know how to reproduce this as it happens inconsistently. I will say that I found something about this on the Internet here: https://stackoverflow.com/questions/14295337/win32gui-setactivewindow-error-the-specified-procedure-could-not-be-found

This is my second time installing dragonfly, and the first time I installed it (probably about 6 months ago), I never ran into this issue. I reinstalled the latest dragonfly two days ago, and now I get this issue a lot when I'm trying to switch to certain programs. It happens when I try to switch to visual studio code, cygwin, and rarely windows explorer, but for whatever reason, I don't notice this exception happen on any of my other apps. I have other `switch to ` commands, but I've never noticed them throw this exception.

The problem seems to go away after a few attempts if I restart the application or manually click to it before trying again. It also seems to randomly work sometimes even if it just failed.

**Comment by tieTYT** [comment]
I don't know if this is just a random fluke that means nothing, but I just said "switch to cygwin" and it printed out "Manage dictation sources". I can't see anyway those 2 things sound alike. I figured I might as well tell you because it might be a clue?
The only reason it might be relevant is because that's the command that often causes the window focus error

**Comment by Danesprite** [comment]
This is happening because the [`Win32Window.set_foreground()`](https://github.com/dictation-toolbox/dragonfly/blob/master/dragonfly/windows/win32_window.py#L228) sometimes doesn't press the control key before calling the Windows `SetForegroundWindow()` function. This is done to satisfy the "the process received the last input event" condition for setting the foreground window ([listed here](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setforegroundwindow#remarks)). Since that condition isn't satisfied, the Python process doesn't have permission to set the window.

I have seen where I messed up and should be able to fix this soon.

**Comment by Danesprite** [comment]
This is now fixed on the master branch. Let me know if the method still fails for you. I hope to release my fix for this soon in dragonfly version 0.25.0.

**Comment by tieTYT** [comment]
thank you for the quick turnaround!

**Comment by LexiconCode** [comment]
Thank you so much Dane.

**Comment by Danesprite** [comment]
No worries :-)

**Comment by LexiconCode** [comment]
@tieTYT you can test with `pip install git+https://github.com/dictation-toolbox/dragonfly.git#egg=dragonfly2`

**Comment by tieTYT** [comment]
okay I ran that command and restarted Dragon. I'll let you know if I run into any problems

**Comment by LexiconCode** [comment]
> 
> 
> okay I ran that command and restarted Dragon. I'll let you know if I run into any problems

I updated the command above it was missing parameter that was needed.

**Comment by tieTYT** [comment]
after I ran this command, i had some interesting side effects. I have a few scripts in `C:\Natlink\Natlink\MacroSystem ` Some of them work, some of them don't. Perhaps this could be an indication of something else broken in this branch?

Here's a script that still works: https://gist.github.com/tieTYT/f87a8583b7c8dfeeaa50ed1efc501b87

Here's a script that doesn't work anymore: https://gist.github.com/tieTYT/7029ef10334fd074e26453a2a5846455

I doesn't work, I mean "those commands are no longer considered commands and print when i say them." It might just be a coincidence, but the 2nd script only should run when Firefox is focused. The first script runs regardless of what's focused.

**Comment by tieTYT** [comment]
okay I tried another script that's in caster that is only supposed to run when visual studio code is focused. Same effect: all commands are printed. I suspect this change made it so that scripts won't ever activate if their activation requires a certain window to be focused?

**Comment by LexiconCode** [comment]
Can you give a breakdown step-by-step using an example of how to reproduce?

1. Step one
2. Step two

**Comment by tieTYT** [comment]
1. Copy that script that doesn't work into your macro system directory.
1. Restart Dragon
1. Switch to Firefox
1. Try to use one of those commands like `new tab `

Expected behavior: Firefox creates a new tab

Actual behavior: Firefox doesn't do anything

**Comment by LexiconCode** [comment]
I believe it's grabbing executable incorrectly from the forefront window. Grabs 'C:' which is the beginning of the absolute path of Firefox.exe.

```
DEBUG:grammar.begin:Grammar g26: detected beginning of utterance.
DEBUG:grammar.begin:Grammar g26: executable 'C:', title 'dragonfly2 · PyPI - Mozilla Firefox'.
DEBUG:context.match:AppContext(['firefox'], None, False): No match, executable doesn't match.
```

**Comment by tieTYT** [comment]
I'm not sure if you're pointing out the change dragonfly needs to make where the change I need to make, but this works fine in .24.

**Comment by LexiconCode** [comment]
> 
> 
> I'm not sure if you're pointing out the change dragonfly needs to make where the change I need to make, but this works fine in .24.

My comment is simply to confirm your issue and to bring a little bit of clarity for Dane. I would like to spend some time to see why this is actually causing an issue but helping out with the wedding tonight.

**Comment by Danesprite** [comment]
Thanks for finding that @LexiconCode. This is happening because of PR #261. I will have that fixed soon.

**Comment by Danesprite** [comment]
The issue with trimming absolute executable paths to "C:" or equivalent is now fixed on the master branch.

**Comment by tieTYT** [comment]
do I use the same command to install it?

**Comment by LexiconCode** [comment]
@tieTYT yes

**Comment by tieTYT** [comment]
Very rarely, I will get this exception:

```action.exec (ERROR): Execution failed: Mimicking failed: No matching rule found for words ['switch', 'to', 'firefox'].
action.exec (ERROR): Execution failed: Mimic('switch', 'to', 'cygwin')+u'npm run compile-twine'+['enter/40']+Mimic('switch', 'to', 'firefox')+Mimic('reload', 'it')
```

But overall, it's working correctly.

**Comment by Danesprite** [comment]
That's good to hear :+1:. I'm not sure why the mimic command is failing. I have noticed that behaviour sometimes. It is probably just a quirk with Dragon.

**Comment by Danesprite** [comment]
I'm closing this now since it is fixed in version 0.25.0.
